lib/elixir/lib/calendar/duration.ex

```diff
@@ -0,0 +1,222 @@
+defmodule Duration do
+  @moduledoc """
+  The Duration type.
+  """
+
+  @default [year: 0, month: 0, week: 0, day: 0, hour: 0, minute: 0, second: 0, microsecond: 0]
```

Comment:
There is a question if we should store microsecond as `{microsecond, precision}`, as in the other calendar types.

Reply:
Yeah i was asking this here https://github.com/elixir-lang/elixir/pull/13385#issuecomment-1975380487 i am not sure tbh. The more i think the about it, the less i think it's relevant on the duration type. Also it's in line with the current functionality we have with `add/3`:

```
NaiveDateTime.add(NaiveDateTime.utc_now(), 100, :microsecond)
```

Reply:
But in `NaiveDateTime.add(NaiveDateTime.utc_now(), 100, :millisecond)` and that would not be possible now.

Reply:
https://github.com/elixir-lang/elixir/pull/13385/commits/75bcb42fe2b2dd383d7e4290f5cef66ef4cb4dcb

It's less flexible and more verbose than supporting `{value, precision}` but i do like that it's explicit and the units all share the same type and format. It's also in line with `NaiveDateTime.add(NaiveDateTime.utc_now(), 100, :millisecond)` but we might as well add support for `nanosecond` if we want to move forward with that approach.

Reply:
^ scrap the above, it really doesn't feel great and adds a lot of bloat for what's already handled more elegantly in the other calendar types. Updated it to this: https://github.com/elixir-lang/elixir/pull/13385/commits/586f4a45af66c15fa80d95f2b2fe8799fc66a9f1

- We take the max precision provided, if any, otherwise we remain the existing precision
- The same is true for adding/subtracting/multiplying durations (btw, should we add `divide/2` as well?)
- Anything else we could do to enforce the format so you can't do: `%Duration{microsecond: 1}`?